src/edu/csus/ecs/pc2/clics/API202306/LanguageService.java

```diff
@@ -0,0 +1,118 @@
+
+    public LanguageService(IInternalContest inContest, IInternalController inController) {
+        super();
+        this.model = inContest;
+        this.controller = inController;
+    }
+
+    /**
+     * Returns a representation of the current model languages in JSON format. The returned value is a JSON array with one language description per array element, complying with 2023-06
+     * 
+     * @param contestId The contest
+     * @return a {@link Response} object containing the model languages in JSON form
+     */
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getLanguages(@PathParam("contestId") String contestId) {
+       
+        // check contest id
+        if(contestId.equals(model.getContestIdentifier()) == false) {
+            return Response.status(Response.Status.NOT_FOUND).build();        
+        }
+        
+        ArrayList<CLICSLanguage> llist = new ArrayList<CLICSLanguage>();
+        
+        // get the languages, one-at-a-time from the model
+        for(Language language: model.getLanguages()) {
+            if (language.isActive()) {
+                llist.add(new CLICSLanguage(language));
+            }
+        }
+        try {
+            ObjectMapper mapper = JSONUtilities.getObjectMapper();
+            String json = mapper.writeValueAsString(llist);
+            return Response.ok(json, MediaType.APPLICATION_JSON).build();
+        } catch (Exception e) {
+            return Response.status(Status.INTERNAL_SERVER_ERROR).entity("Error creating JSON for languages " + e.getMessage()).build();
+        }
+    }
+
+    /**
+     * Returns a representation of the specified language for the specified contest in JSON format. The returned value is compliant with 2023-06
+     * 
+     * @param contestId The contest
+     * @param languageId The language
+     * @return
+     */
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    @Path("{languageId}/")
+    public Response getLanguage(@PathParam("contestId") String contestId, @PathParam("languageId") String languageId) {
+        
+        // get the languages, one-at-a-time from the model
+        for(Language language: model.getLanguages()) {
+            if (language.isActive() && language.getElementId().toString().equals(languageId)) {
+                return Response.ok(new CLICSLanguage(language).toJSON(), MediaType.APPLICATION_JSON).build();
+            }
+        }
+        return Response.status(Response.Status.NOT_FOUND).build();
+    }
+    
+    /**
+     * Retrieve access information about this endpoint for the supplied user's security context
+     * 
+     * @param sc User's security information
+     * @return CLICSEndpoint object if the user can access this endpoint's properties, null otherwise
+     */
+    public static CLICSEndpoint getEndpointProperties(SecurityContext sc) {
+        return(new CLICSEndpoint("languages", JSONUtilities.getJsonProperties(CLICSLanguage.class)));
```

Comment:
I guess I must not understand how the whole `getEndpointProperties()` thing is working, because here again I see a `SecurityContext` parameter and multiple comment lines (above) that claim that endpoing access is being controlled according to that `SecurityContext` -- but the `sc` parameter is never used.

Reply:
See my response to the `JudgementTypes `question.  In a nutshell, the `getEndpointProperties()` method is called by `CLICSContestAccess.invokeServiceEndpointAccessMethod` which has no idea which class's `getEndpointProperities() `is being called, or whether it needs the `sc` or not, so it is always passed and left up to the particular class's implementation to decide if it needs the `sc `or not.